---
title: "R Notebook"
output: 
---

```{r echo = F}
library(data.table)
library(GenomicRanges)
```

```{r}
gtex <- fread('src/gtex/junctions_1to22XY_gtexv2_filt.tsv.gz')

sra <- fread('src/sra/junctions_1to22XY_srav3_filt.tsv.gz')
```


# step 1: merge sra & gtex sj data
	- Columns to join on: chrom, start, end, strand
	- Columns to add: sra_samples_count, gtex_samples_count, gtex_max_reads


```{r}
gtex <- gtex[, .(chrom, start, end, strand, sample_count)]
sra <- sra[, .(chrom, start, end, strand, sample_count)]
```


```{r}
merge <- merge(gtex, sra, by = c('chrom', 'start', 'end', 'strand'), 
               all = TRUE, suffixes = c('_gtex', '_sra'))
merge[, chrom := gsub('chr', '', chrom)]
```



# step 2: Filter to splice junctions where 1 end is annotated
```{r}
gtf <- rtracklayer::import('ref/Homo_sapiens.GRCh38.110.chr.gtf.gz')
gtf_dt = as.data.table(gtf)
exons = gtf_dt[type == 'exon', .(seqnames = transcript_id, start, end, strand)]
transcripts = gtf_dt[type == 'transcript', .(seqnames = transcript_id, start, end, strand)]
introns = GenomicRanges::setdiff(GRanges(transcripts), GRanges(exons))
introns_dt = as.data.table(introns)
 
introns_dt= gtf_dt[type == 'transcript', .(seqnames = transcript_id, chrom = seqnames)][introns_dt, on = .(seqnames)]
```


```{r}

# unique annotated starts
ann_starts <- unique(introns_dt[, .(chrom, start)])
ann_starts[, annotated_start := 1]

# unique annotated ends
ann_ends <- unique(introns_dt[, .(chrom, end)])
ann_ends[, annotated_end := 1]

# unique annotated sjs
ann_sjs <- unique(introns_dt[, .(chrom, start, end)])
ann_sjs[, annotated_sj := 1]

```


```{r}
merge <- ann_starts[merge, on = .(chrom, start)]
merge <- ann_ends[merge, on = .(chrom, end)]
merge <- ann_sjs[merge, on = .(chrom, start, end)]
merge[is.na(merge)] <- 0
mergefilt <- merge[annotated_start == 1 | annotated_end == 1]
```




# step 3: annotate with basic mis-splicing event categories

```{r}
matrix <- data.table(strand = c(), annotated_intron = c(), annotated_start = c(), annotated_end = c(), splicing_event_class = c())
matrix[, strand := rep(c('+', '-'), 4)]
matrix[, annotated_sj := c(1,1,rep(0, 6))]
matrix[, annotated_start := c(1,1,1,1,1,0,0,1)]
matrix[, annotated_end := c(1,1,1,1,0,1,1,0)]
matrix[, splicing_event_class := c(rep('normal splicing',2), rep('exon skipping', 2), rep('cryptic acceptor', 2), rep('cryptic donor', 2))]

mergefilt <- matrix[mergefilt, on = .(strand, annotated_sj, annotated_start, annotated_end)]
```


```{r}
mergefilt[strand == '+' & annotated_start == 1, annotated_donor := 1]
mergefilt[strand == '-' & annotated_end == 1, annotated_donor := 1]


mergefilt[strand == '+' & annotated_end == 1, annotated_acceptor := 1]
mergefilt[strand == '-' & annotated_start == 1, annotated_acceptor := 1]

mergefilt[is.na(annotated_donor), annotated_donor := 0]
mergefilt[is.na(annotated_acceptor), annotated_acceptor := 0]
```

```{r}
mergefilt[, sample_count := sample_count_gtex + sample_count_sra]
```

# save
```{r}
fwrite(mergefilt[, .(chrom, start, end, strand, annotated_donor, annotated_acceptor, 
                 splicing_event_class, sample_count_gtex, sample_count_sra, sample_count)], 'src/junctions_merged_processed.tsv.gz')
```




